index.d.ts

```diff
@@ -1,288 +1,591 @@
+  manualSorting: boolean
+  disableSorting: boolean
+  disableMultiSort: boolean
+  isMultiSortEvent: (e: MouseEvent) => boolean
+  maxMultiSortColCount: number
+  disableSortRemove: boolean
+  disabledMultiRemove: boolean
+  orderByFn: (
+    rows: Row<D>[],
+    sortFns: SortByFn<D>[],
+    directions: boolean[]
+  ) => Row<D>[] // CHECK
+  sortTypes: Record<string, SortByFn<D>>
+}>
+
+export interface UseSortByHooks<D extends object> {
+  getSortByToggleProps: ((
+    column: Column<D>,
+    instance: TableInstance<D>
+  ) => object)[]
+}
 
-  export function useSortBy<D = {}>(
-    props: TableOptions<D>
-  ): TableOptions<D> & {
-    rows: Row<D>[]
-  }
+export interface UseSortByState<D extends object> {
+  sortBy: SortingRule<D>[]
+}
 
-  export function useGroupBy<D = {}>(
-    props: TableOptions<D>
-  ): TableOptions<D> & { rows: Row<D>[] }
+export type UseSortByColumnOptions<D extends object> = Partial<{
+  disableSorting: boolean
+  sortDescFirst: boolean
+  sortInverted: boolean
+  sortType: SortByFn<D> | DefaultSortTypes | string // TODO: Check this works
+}>
+
+export interface UseSortByInstanceProps<D extends object> {
+  rows: Row<D>[]
+  preSortedRows: Row<D>[]
+  toggleSortBy: (
+    columnId: IdType<D>,
+    descending: boolean,
+    isMulti: boolean
+  ) => void
+}
 
-  export function usePagination<D = {}>(
-    props: TableOptions<D>
-  ): UsePaginationValues<D>
+export interface UseSortByColumnProps<D extends object> {
+  canSort: boolean
+  toggleSortBy: (descending: boolean, multi: boolean) => void
+  getSortByToggleProps: (props?: object) => object
+  clearSorting: () => void
+  isSorted: boolean
+  sortedIndex: number
+  isSortedDesc: boolean | undefined
+}
 
-  export function useExpanded<D = {}>(
-    props: TableOptions<D>
-  ): TableOptions<D> & {
-    toggleExpandedByPath: () => any
-    expandedDepth: []
-    rows: Row<D>[]
+export type SortByFn<D extends object> = (
+  rowA: Row<D>,
+  rowB: Row<D>,
+  columnId: IdType<D>
+) => 0 | 1 | -1
+
+export type DefaultSortTypes = 'alphanumeric' | 'datetime' | 'basic'
+
+export interface SortingRule<D> {
+  id: IdType<D>
+  desc?: boolean
+}
+/* #endregion */
+
+// Additional API
+export function useTableState<D extends object = {}>(
+  initialState?: Partial<TableState<D>>,
+  overrides?: Partial<TableState<D>>,
+  options?: {
+    reducer?: (
+      oldState: TableState<D>,
+      newState: TableState<D>,
+      type: string
+    ) => TableState<D>
+    useState?: typeof useState
   }
+): TableStateTuple<D>
 
-  export function useTableState<D = {}>(
-    initialState?: Partial<TableState<D>>,
-    overriddenState?: Partial<TableState<D>>,
-    options?: {
-      reducer?: (
-        oldState: TableState<D>,
-        newState: TableState<D>,
-        type: string
-      ) => any
-      useState?: typeof useState
-    }
-  ): State<D>
+export const actions: Record<string, string>
+export function addActions(...actions: string[]): void
```

Comment:
can we add:

```tsx
  export const defaultState: any
```

Reply:
I removed it. It has no value publicly, AFAIK. Also, `defaultState` would be `{}` or some kind of `Record` seeing as that what is defined in the code.

Reply:
It has value publicly if you are writing your own hooks.